Code review for getting the Poll/Survey XBlocks on edx.org

antoviaque opened the pull request:
Fake PR for review purposes. Do not merge.

Note: ignore the first comments by mtyaka - these were from a previous PR where comments were made directly on the commits instead of on the PR.

Kelketek committed:
First Prototype.

Kelketek committed:
CMS editing implemented but not working quite right yet.

Kelketek committed:
Polished up backend. Markdown needs extraction. Frontend needs further CSS.

Kelketek committed:
Semantic correction to answer adding button id name.

Kelketek committed:
Removed Pen.js. It was nice, but too buggy. Using normal textarea instead.

Kelketek committed:
Fixed a bug where the tally wasn't being updated.

Kelketek committed:
Added markdown to the requirements.txt.

Kelketek committed:
Began work on formatting the results per the specification.

Kelketek committed:
Fixed CSS for display of results.

Kelketek committed:
CSS for results should match specification.

Kelketek committed:
Several CSS fixes. Dropped title field. Fixed markdown handling. Added Feedback.

Kelketek committed:
Further refinement and test stubs added.

Kelketek committed:
Tests written.

Kelketek committed:
Tightened up tests, added image support, question reordering.

Kelketek committed:
Markup and CSS fixes, poll form scoping.

Kelketek committed:
Removed static files for Pen.js.

Kelketek committed:
Several style fixes per review feedback.

Kelketek committed:
Better handling of JSON.

antoviaque committed:
Merge pull request #1 from open-craft/first-version

(WIP) First Version.

Kelketek committed:
Removed several security features

CMS studio users will be considered trusted from now on. Unfiltered HTML
will be permitted alongside markdown. This does mean that privilege
escalation is easy, as a Javascript snippet could be used to grant
superuser privileges if a superuser visits a page.

Kelketek committed:
Presentation tweaks and scaffolding for Survey.

Kelketek committed:
Refactored JS and Python classes for better code reuse.

Kelketek committed:
Test and Javascript refactoring.

Kelketek committed:
Vote functionality implemented for Survey.

Kelketek committed:
Tests added to Survey.

Kelketek committed:
CSS Stylings for Survey.

Kelketek committed:
Several bits of refactoring while preparing for Survey studio view.

Kelketek committed:
Studio view and images implemented for Surveys.

Kelketek committed:
More markdown areas, scroll handling, README, tests.

Kelketek committed:
Fix tally cleaning, poll scenario, poll result label CSS.

Kelketek committed:
Update README.md

Kelketek committed:
Improve event data and response display for surveys.

Kelketek committed:
Added example of analytics events in the README.

Kelketek committed:
Add 'Feedback' header to feedback.

antoviaque committed:
Merge pull request #2 from open-craft/survey

Version 2 with Survey

Kelketek committed:
Several appearance tweaks.

Kelketek committed:
Fixed several issues with studio authoring.

antoviaque committed:
Merge pull request #1 from open-craft/fixes

Addressed bugs from QA

Kelketek committed:
Appearance tweaks, per QA Feedback.

Kelketek committed:
Added studio tests.

Kelketek committed:
Use studio_view argument with Xblock-utils.

Also make one of the tests that used Python literals use JSON while XBlock is in flux.

Kelketek committed:
Fix issue that prevented poll submission on first creation in studio.

Kelketek committed:
New and Improved documentation

antoviaque committed:
Merge pull request #2 from open-craft/style_tweaks

Appearance tweaks and studio tests, per QA Feedback.

antoviaque committed:
Merge pull request #3 from open-craft/doc-updates

Doc updates

Kelketek committed:
Added ability to have polls that don't share results with users.

Kelketek committed:
Merge pull request #4 from open-craft/private_results

Added ability to have polls that don't share results with users.

antoviaque committed:
Update licensing - Release the XBlock under the AGPLv3 license

zvishal committed:
Merge pull request #5 from open-craft/license-agpl

Update licensing - Release the XBlock under the AGPLv3 license

Kelketek committed:
Added 'max submissions' functionality.

Kelketek committed:
Remove unused utils and associated tests.

mtyaka committed:
Tweaks for pep8/pylint.

Kelketek committed:
Addressed review notes.

mtyaka committed:
Add ability to grant view results permission to additional groups.

You can now use the XBLOCK_POLL_EXTRA_VIEW_GROUPS django setting
to allow users from additional groups to view the results.

By default only course staff has permission to view the results.

antoviaque committed:
Merge pull request #6 from mckinseyacademy/max_attempts

Max Submissions

Kelketek committed:
Addressed nits from max_submissions task.

Kelketek committed:
Added counter display for polls with maximum submissions greater than 1.

antoviaque committed:
Merge pull request #8 from mckinseyacademy/cleanup

Addressed nits and remaining work from max submissions task.

Kelketek committed:
Fixed Javascript scoping issues and missing return statement in vote.

mtyaka committed:
Merge pull request #9 from mckinseyacademy/cleanup

Addressed remaining issues from second review

mtyaka committed:
Simplify test setup and set up travis CI.

No need to manually install xblock-sdk & dependencies anymore.
Inside a fresh virtual environment, run `sh install_test_deps.sh`,
then run the tests with `python run_tests.py`. That's it.

mtyaka committed:
Add ability for staff to always view the results.

Course staff users can now view the results without voting.
The can also see the results for polls/surveys marked as 'private'.

If user has sufficient permissions, a 'View results' link is shown
on the bottom of the component in the LMS.

mtyaka committed:
Merge pull request #10 from open-craft/show-private-results

Add ability for course staff to always view the results

Kelketek committed:
Allow single question surveys (and single answer polls, for funzies)

Kelketek committed:
Merge pull request #11 from mckinseyacademy/allow_one

Allow single question surveys (and single answer polls, for funzies)

Kelketek committed:
More thorough cleaning of Survey Tally.

Kelketek committed:
Merge pull request #12 from mckinseyacademy/survey_tally_clean

More thorough cleaning of Survey Tally.

antoviaque commented:
First review comments from @sarina:

> Looking at it _super_ briefly, I think the chief concerns will likely be accessibility (for example, orange text is likely a no-no).
> Also we may want to change the readme to provide best practices (for example, images without text should be avoided, or provide an alt-text somehow).

sarina commented:
@antoviaque I'm going to spin up a sandbox with this block installed. Do you have any documentation on installation/studio use/lms use?

sarina commented:
n/m I see the readme, I've already read the readme, I should have had more coffee today

sarina commented on README.md:
```diff
@@ -0,0 +1,239 @@
+Here is an example of a poll that uses markdown in the question and the answers, and which uses images for each
+of the answers:
+
+![Image-only poll](doc_img/img_poll.png)
```
I'm concerned about this from an a11y point of view. I'll let Mark comment more but want to flag this. Images-only will need to have alt text. Alternatively, we could say, these aren't accessible and you shouldn't use them or something.

sarina commented on README.md:
```diff
@@ -0,0 +1,239 @@
+
+![Survey results](doc_img/survey_result.png)
+
+The top choice's percentage is shown in *orange* while the user's selection is marked with a shaded background.
```
I'm concerned about the color choices here.

Also the table is really squashed. Can we add more padding to the cells and especially add height padding?

sarina commented on poll/public/html/poll_edit.html:
```diff
@@ -0,0 +1,96 @@
+                </select>
+            </div>
+            <span class="tip setting-help">
+                If this is set True, don't display results of the poll to the user.
```
"If this is set to True"

smarnach replied:
Done.

sarina commented on poll/public/css/poll.css:
```diff
@@ -0,0 +1,218 @@
+    padding: .5em;
+    display: inline-block;
+    background-color: #e5ebee;
+    color: #e37222;
```
I'm concerned about the contrast here

e-kolpakov replied:
@sarina I believe this is best addressed using customizable themes, like in mentoring/problem_builder:
- Obtaining theme configuration: https://github.com/open-craft/problem-builder/blob/master/problem_builder/mentoring.py#L208-219
- Adding a theme to rendered fragment: https://github.com/open-craft/problem-builder/blob/master/problem_builder/mentoring.py#L268-272

Default theme will than contain proper (from a11y point of view) colors, and there would be an alternative theme containing this particular config, as it is used by third party client.

sarina replied:
:+1: this makes sense.

sarina commented on README.md:
```diff
@@ -0,0 +1,239 @@
+
+**Notes on Private Results**: Users will be able to change their vote on polls and surveys with this option enabled.
+An analytics event will not be fired upon the student viewing the results, as the results are never visible. A user
+will see a thank you message and any feedback provided upon submission:
```
The screenshot that follows this is a bit concerning. I'd change this line to read, for clarity:

A user will see a thank you message, and optionally, any instructor-provided Feedback in an additional "Feedback" section, when they click submit.

smarnach replied:
Done.

sarina commented on poll/poll.py:
```diff
@@ -0,0 +1,849 @@
+        """
+        if self.max_submissions == 0:
+            return True
+        if self.max_submissions > self.submissions_count:
```
why not

``` python
if 0 <= self.max_submissions < self.submissions_count
```

e-kolpakov replied:
@sarina `max_submissions == 0` means that there are no limit for resubmitting. Hence, `True` should be returned if either is true:
- max submission is zero
- `max_submisions` is **greater** than `submissions_count`.

smarnach replied:
I've changed this to

```
return self.max_submissions == 0 or self.submissions_count < self.max_submissions
```

which is both more concise and clearer in my opinion.

sarina commented on poll/public/css/poll.css:
```diff
@@ -0,0 +1,218 @@
+    max-width: 80%;
+}
+
+li.poll-answer {
```
maybe my css is just bad but isn't the proper way to scope this to be

``` css
.poll-answer li {
    display: block;
}
```

?

e-kolpakov replied:
@sarina `li.poll-answer` means "apply to `li` elements that have class `poll-answer`. `.po;;-answer li` means "apply to `li` elements that are _children_ of any element that have class `poll-answer`. So it's not quite the same.

sarina commented on poll/public/html/poll_edit.html:
```diff
@@ -0,0 +1,96 @@
+            </div>
+            <span class="tip setting-help">
+                Maximum number of times a user may submit a poll. <strong>Setting this to a value other than 1 will imply that
+                'Private Results' should be true.</strong> Setting it to 0 will allow infinite resubmissions.
```
? Why does setting this to 1 imply private results?

e-kolpakov replied:
@Kelketek if "Private Results" is set to False and **more than one** resubmissions allowed, students can vote, review other's results and than change their mind. Second vote will be biased by the current state of voting and may result in statistically incorrect poll results. Hence, it is required to set "Private Results" to True when "Maximum Submissions" is _not_ one.

Kelketek replied:
I'm guessing I'm being pinged to confirm this? @e-kolpakov 's explanation is correct.

sarina commented:
@antoviaque is it possible that you guys have any test .tar.gz courses with lots of examples of the various types of poll blocks?

antoviaque commented:
@sarina Thanks for setting up the sandbox! @e-kolpakov might still need to setup one on our sandboxes to be able to update with his changes, especially the ones related to theming, but it's helpful to have it already. 

For examples, we do have courses that contain examples, but they are the client's proprietary content and we can't put them on a public instance. But there aren't a lot of different configurations though - @e-kolpakov do you think you would have time to setup examples of the different usage scenarios on the sandbox? If not that might also be a good follow-up task for next week's sprint, which @Kelketek could take on since he did most of the original implementation. Let me know.

e-kolpakov commented:
@sarina Here's how it looks like with LMS theme applied: 

![poll lms theme](https://cloud.githubusercontent.com/assets/3330048/7589612/5683232a-f8cc-11e4-8003-00f05af323ff.png)

I'm sending a PR shortly and probably will have a sandbox showcasing this available soon.

e-kolpakov commented:
@sarina sandbox with a course showcasing some features are there: http://sandbox3.opencraft.com/courses/TestOrg/ImportCourse1/now

e-kolpakov commented:
@sarina requested changes are implemented at https://github.com/open-craft/xblock-poll/pull/2. Waiting for a11y review.

Also, does it make sense to update documentation screenshots? The course and images used to create them were lost, so it would take a bit of time to recreate them. The only difference would be css theme though.

sarina commented:
>  does it make sense to update documentation screenshots? The course and images used to create them were lost, so it would take a bit of time to recreate them. The only difference would be css theme though.

Updated screenshots would make it easier for people to assess whether or not they want to use the block. I would hate for people to not use the block because they don't like the way it looks, or vice versa.

I'll follow up with Mark about the state of a11y review (ticket here: [AC-96](https://openedx.atlassian.net/browse/AC-96))

sarina commented:
Also we're going to want some "modern" screenshots to publicize the work, once it's available.

cptvitamin commented on poll/public/handlebars/poll_results.handlebars:
```diff
@@ -0,0 +1,43 @@
+    {{#each tally}}
+        <li class="poll-result">
+            <div class="poll-result-input-container">
+              <input id="answer-{{key}}" type="radio" disabled {{#if choice}}checked="True"{{/if}} />
```
`"True"` is not a valid value for the checked attribute (boolean)
http://www.w3.org/html/wg/drafts/html/CR/infrastructure.html#boolean-attribute

It's either `checked="checked"` or simply `checked` for checked attributes and omitting the `checked` attribute all together if its not checked.

smarnach replied:
Done.

sarina replied:
Where can the updated code be viewed? When should I take another look?
Finally, let me know when the sandbox is updated. Thanks :)

On Thu, Jun 4, 2015 at 3:28 PM, Sven Marnach notifications@github.com
wrote:

> In poll/public/handlebars/poll_results.handlebars
> https://github.com/mckinseyacademy/xblock-poll/pull/13#discussion_r31756290
> :
> 
> > @@ -0,0 +1,43 @@
> > +<script id="poll-results-template" type="text/html">
> > -    <h3 class="poll-header">{{display_name}}</h3>
> > -    <div class="poll-question-container">{{{question}}}</div>
> > -    <ul class="poll-answers-results poll-results">
> > -    {{#each tally}}
> > -        <li class="poll-result">
> > -            <div class="poll-result-input-container">
> > -              <input id="answer-{{key}}" type="radio" disabled {{#if choice}}checked="True"{{/if}} />
> 
> Done.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/mckinseyacademy/xblock-poll/pull/13/files#r31756290.

smarnach replied:
@sarina Sorry, the comment "Done" is primarily addressed to myself, to keep track of what has already been addressed and what's still open.  There are still quite a few points open.  When I'm done with everything, I'll update the sandbox and let you know.

sarina replied:
Sven - thanks for the update! That's exactly what I meant to clarify. :)

On Thu, Jun 4, 2015 at 5:42 PM, Sven Marnach notifications@github.com
wrote:

> In poll/public/handlebars/poll_results.handlebars
> https://github.com/mckinseyacademy/xblock-poll/pull/13#discussion_r31768930
> :
> 
> > @@ -0,0 +1,43 @@
> > +<script id="poll-results-template" type="text/html">
> > -    <h3 class="poll-header">{{display_name}}</h3>
> > -    <div class="poll-question-container">{{{question}}}</div>
> > -    <ul class="poll-answers-results poll-results">
> > -    {{#each tally}}
> > -        <li class="poll-result">
> > -            <div class="poll-result-input-container">
> > -              <input id="answer-{{key}}" type="radio" disabled {{#if choice}}checked="True"{{/if}} />
> 
> @sarina https://github.com/sarina Sorry, the comment "Done" is
> primarily addressed to myself, to keep track of what has already been
> addressed and what's still open. There are still quite a few points open.
> When I'm done with everything, I'll update the sandbox and let you know.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/mckinseyacademy/xblock-poll/pull/13/files#r31768930.

cptvitamin commented on poll/public/handlebars/poll_results.handlebars:
```diff
@@ -0,0 +1,43 @@
+                <div class="poll-image result-image">
+                    <label for="answer-{{key}}" class="poll-image-label">
+                        {{#if img}}
+                          <img src="{{img}}" />
```
All images _MUST_ have an `alt` attribute (these images do not).  While it is acceptable in some contexts to have an null alt (`alt=""`) this would not be one of them seeing as though some description of the image would be required to differentiate one choice from the other.
http://www.w3.org/html/wg/drafts/html/CR/embedded-content-0.html#alt

There must be a mechanism for the poll's author to provide a description for each image.  I recommend adding text that helps the author understand the meaning of useful alternatives (i.e. "Describe this image in a way that would allow someone to answer the poll if the image did not load")

smarnach replied:
Done.

cptvitamin replied:
@smarnach when I look at the example in the sandbox, I see that alt attribute value as "None".  This isn't going to work.  If a user cannot see the image, how are they to differentiate between one image and the next?  Which is the cutest cat?  None, none, none or none.

sarina replied:
@cptvitamin Just for your reference, the responses to comments on this pr are (unfortunately) on another PR: https://github.com/open-craft/xblock-poll/pull/2

smarnach replied:
@cptvitamin Thanks for pointing this out -- this was a bug in one of the Django templates, resulting in `None` to be shown instead of the empty string.  (We still have to show an empty string if no alternate description is available, e.g. for block usages that have been created before it was possible to add alt tags.)

sarina replied:
@smarnach would it be reasonable/possible to display an empty string for backwards compatibility of images that don't have alt defined, but enforce for any new or edited problems that alt attributes must be added?

smarnach replied:
@sarina That would be possible, but I feel it would be a bit weird to force people who change something unrelated to add the alt descriptions.  Moreover, this XBlock is also used outside of edx.org, in contexts that have less strict accessibility requirements.  And finally, as stated before, I don't think people will enter anything useful just because they are forced to enter _something_.

cptvitamin commented on poll/public/handlebars/poll_results.handlebars:
```diff
@@ -0,0 +1,43 @@
+            {{/if}}
+            <div class="percentage-gauge-container">
+              <div class="percentage-gauge" style="width:{{percent}}%;">
+                  <label class="poll-answer-label" for="answer-{{key}}">{{{answer}}}</label>
```
the value of `{{{answer}}}` seems to include presentation markup (`<p>` elements in the example I reviewed).  Only [phrasing content](http://www.w3.org/html/wg/drafts/html/CR/dom.html#phrasing-content-1 is allowed as a child fo `<label>` (`<p>`'s are not allowed).  I would just include plain text here without any other markup.

smarnach replied:
We allow to use Markdown in the answers, mainly to make it easy to add emphasis and links (well, to add phrasing content).  Markdown wraps all text in `<p>` elements, and this is how they end up in the label.

This is rather tricky to fix.  We can't turn the label into anything else, since it needs to be associated with the radio button, and we can't turn off Markdown without breaking backwards compatibility.  We might be able to filter out tags that are not phrasing content, but that would be a bit of effort to get done.

How important is it to get this fixed?  While this is invalid HTML, existing browsers don't seem to have a problem with it.  Will it be a problem for screen readers?

cptvitamin replied:
@smarnach I don't think paragraph elements in a <label> are the end of the world, if we knew for sure that would be the worst thing that was ever put in there.  When I think of everything we can do with Markdown, I get a little nervous.  Multiple paragraphs would be a problem.  Links would be a problem (even though they qualify as phrasing content).  

Personally, I wouldn't write invalid HTML, but if you can sleep at night, I'm OK with it ;).  I'm convinced that this is something we can address with documentation (by highlighting the features of Markdown that are allowed here, really just italics and bold).

smarnach replied:
@cptvitamin I don't want to write invalid HTML either, but as pointed out above, I don't see how to fix this without breaking backwards compatibility.  I've added a hint pointing out that you should only use bold or italics here.

cptvitamin commented on poll/public/handlebars/poll_results.handlebars:
```diff
@@ -0,0 +1,43 @@
+            </div>
+        </li>
+       {{^last}}
+        <li class="poll-spacer">
```
Please use CSS to create "space."  A screen reader will be told that there is one more option than there really is when they encounter this list.

smarnach replied:
It would be nice if that would be possible in all cases.  This is using `display: table`, and it turns out that CSS doesn't allow to add space between table rows – margin and border are simply ignored at least by Webkit browsers.  I tried recreating the same layout with `display: inline-block`, but there are many things I can't even get to work on a single browser the same way it was before, let alone on all supported browsers.

How bad is this for screen readers?  It will likely be at least half a day's work to get this fixed properly.

cptvitamin replied:
@smarnach It's actually pretty bad for screen reader users.  I just played around with the CSS and found that you can use `border-bottom: 4px solid transparent;`  on the `<li>` 's to achieve the same affect you are getting with the blank `<li>`

smarnach replied:
@cptvitamin This was the second thing (of many) I tried, but unfortunately it works on neither Chrome nor Firefox.  (Just curious: what browser did this you get this to work on?)

If this is pretty bad for screen readers, we have to find a different solution.

@antoviaque Do you have any suggestions how to proceed?  Simply removing the spacers here will change the style for McKinsey, which we probably don't want to do, and recreating the same layout we currently have on all browsers using an approach that doesn't need the spacers is a lot of work and likely to introduce styling incompatibilities and bugs.

cptvitamin commented on poll/public/handlebars/poll_studio.handlebars:
```diff
@@ -0,0 +1,22 @@
+{{#each items}}
+<li class="field comp-setting-entry is-set poll-{{noun}}-studio-item">
+    <div class="wrapper-comp-setting">
+        <label class="label setting-label poll-setting-label" for="{{noun}}-label-{{key}}">{{noun}}</label>
```
if you make the associated `<input>` a child of the `<label>` you don't have to worry about the `for/id` references.

smarnach replied:
Thanks for the tip!  I'll leave it the way it is here, since it also works.

cptvitamin commented on poll/public/handlebars/poll_studio.handlebars:
```diff
@@ -0,0 +1,22 @@
+        <input class="input setting-input" name="{{noun}}-label-{{key}}" id="{{noun}}-label-{{key}}" value="{{text}}" type="text" /><br />
+        {{#if image}}
+        <label class="label setting-label" for="{{noun}}-img-{{key}}">Image URL</label>
+        <input class="input setting-input" name="{{noun}}-img-{{key}}" id="{{noun}}-img-{{key}}" value="{{img}}" type="text" />
```
There needs to be a mechanism for describing the image here.  This is imperative.  A description of the image would be required to differentiate one choice from the other for users who cannot see.
http://www.w3.org/html/wg/drafts/html/CR/embedded-content-0.html#alt

I also recommend adding text that helps the author understand the meaning of useful alternatives (i.e. "Describe this image in a way that would allow someone to answer the poll if the image did not load")

smarnach replied:
Done.

cptvitamin commented on poll/public/handlebars/poll_studio.handlebars:
```diff
@@ -0,0 +1,22 @@
+        <input class="input setting-input" name="{{noun}}-img-{{key}}" id="{{noun}}-img-{{key}}" value="{{img}}" type="text" />
+        {{/if}}
+        <div class="poll-move">
+            <div class="poll-move-up">&#9650;</div>
```
This control is inaccessible (`<div>` elements have no accessible roles, states or properties associated with them)  The user cannot move focus with the keyboard, access an accessible name, or otherwise interact with this control at all.  Recommend the following:
`<button class="poll-move-up">&#9650;<span class="sr">&nbsp;move poll up</span></button>`
Button elements come with all sorts of cool behaviors and properties for free, like you can focus on them, click them, get their accessible name, etc.  Same applies for the move down control below.

smarnach replied:
Done.

cptvitamin commented on poll/public/handlebars/poll_studio.handlebars:
```diff
@@ -0,0 +1,22 @@
+    <span class="tip setting-help">
+        {{#if image}}This must have an image URL or text, and can have both.{{/if}}
+    </span>
+    <a href="#" class="button action-button poll-delete-answer" onclick="return false;">Delete</a>
```
This should be a `<button>` not an `<a>`.  Buttons do something and links go somewhere.  If you find yourself with an (invalid) URL ref of `#` it usually means you should be using a button.

smarnach replied:
Done.

cptvitamin commented on poll/public/handlebars/poll_studio.handlebars:
```diff
@@ -0,0 +1,22 @@
+        </div>
+    </div>
+    <span class="tip setting-help">
+        {{#if image}}This must have an image URL or text, and can have both.{{/if}}
```
In addition to there being a mechanism for providing an alt attribute, I recommend making it a required form field if the image field is not null to enforce the inclusion of an alt attribute for every image.

smarnach replied:
I added a form field and a comment to strongly encourage filling it in, but didn't go all the way to make it mandatory to support the legacy content that doesn't have any alt texts applied.  We can't verify whether people entered anything useful anyway.

cptvitamin commented on poll/public/handlebars/survey_results.handlebars:
```diff
@@ -0,0 +1,37 @@
+                <td class="survey-question">
+                    {{#if img}}
+                        <div class="poll-image-td">
+                            <img src="{{img}}" />
```
meaningful `alt` attribute required.

smarnach replied:
Done.

cptvitamin commented on poll/public/html/poll.html:
```diff
@@ -0,0 +1,48 @@
+      <div class="view-results-button-wrapper"><a class="view-results-button">View results</a></div>
+  {% endif %}
+{% endif %}
+</div>
```
Please see feedback on .handlebars templates

smarnach replied:
Done.

cptvitamin commented on poll/public/html/poll_edit.html:
```diff
@@ -0,0 +1,96 @@
+    <ul class="list-input settings-list" id="poll-line-items">
+        <li class="field comp-setting-entry is-set">
+            <div class="wrapper-comp-setting">
+                <label class="label setting-label poll-setting-label" for="poll-display-name">Display Name</label>
```
for future reference, if you make the associated `<input>` a child of the `<label>` you don't have to worry about the `for/id` references.

smarnach replied:
Acknowledged.

cptvitamin commented on poll/public/html/poll_edit.html:
```diff
@@ -0,0 +1,96 @@
+              <h2><label for="poll-question-editor">Question/Prompt</label></h2>
+              <a href="//daringfireball.net/projects/markdown/syntax" target="_blank">Markdown Syntax</a> is supported.
+              <div id="poll-question-editor-container">
+                <textarea class="input setting-input" name="question" id="poll-question-editor">{{question}}</textarea>
```
Please put an `id` on the "tip" `<span>` below and reference it using an `aria-describedby="[ID]"` attribute on this form field

smarnach replied:
Done.

cptvitamin commented on poll/public/html/poll_edit.html:
```diff
@@ -0,0 +1,96 @@
+           <h2><label for="poll-feedback-editor">Feedback</label></h2>
+           <a href="//daringfireball.net/projects/markdown/syntax" target="_blank">Markdown Syntax</a> is supported.
+           <div id="poll-feedback-editor-container">
+             <textarea class="input setting-input" name="feedback" id="poll-feedback-editor">{{feedback}}</textarea>
```
Please put an `id` on the "tip" `<span>` below and reference it using an `aria-describedby="[ID]"` attribute on this form field

smarnach replied:
Done.

cptvitamin commented on poll/public/html/poll_edit.html:
```diff
@@ -0,0 +1,96 @@
+        <li class="field comp-setting-entry is-set">
+            <div class="wrapper-comp-setting">
+                <label class="label setting-label poll-setting-label" for="poll-private-results">Private Results</label>
+                <select id="poll-private-results" class="input setting-input" name="private_results">
```
Please put an `id` on the "tip" `<span>` below and reference it using an `aria-describedby="[ID]"` attribute on this form field

smarnach replied:
Done.

cptvitamin commented on poll/public/html/poll_edit.html:
```diff
@@ -0,0 +1,96 @@
+        <li class="field comp-setting-entry is-set">
+            <div class="wrapper-comp-setting">
+                <label class="label setting-label poll-setting-label" for="poll-max-submissions">Maximum Submissions</label>
+                <input id="poll-max-submissions" type="number" min="0" step="1" value="{{ max_submissions }}"/>
```
Please put an `id` on the "tip" `<span>` below and reference it using an `aria-describedby="[ID]"` attribute on this form field

smarnach replied:
Done.

cptvitamin commented on poll/public/html/poll_edit.html:
```diff
@@ -0,0 +1,96 @@
+    <div class="xblock-actions">
+        <ul>
+            <li class="action-item" id="poll-add-answer">
+                <a href="#" class="button action-button" class="poll-add-item-link" onclick="return false;">Add Answer</a>
```
This should be a `<button>` not an `<a>`.  Buttons do something and links go somewhere.  If you find yourself with an (invalid) URL ref of `#` it usually means you should be using a button.  Apply to all buttons below.

smarnach replied:
The buttons at the bottom of the edit dialogue are a links for all component types in edX.  If this gets changed in edX itself, I think it makes sense to change it here as well, but it would be weird if this XBlock would be the only exception.

cptvitamin commented on poll/public/html/poll_edit.html:
```diff
@@ -0,0 +1,96 @@
+                </p>
+            {% endif %}
+        </li>
+        <li id="poll-answer-marker"></li>
```
If these are not intended to have a value, they should not be `<li>`'s

smarnach replied:
The only tags allowed inside `ul` are `li`, `ol` and `ul`.  Would empty `ul`s be better?

smarnach replied:
We will redesign this UI anyway in a separate PR, moving questions and answers to a different tab in the dialogue.  This redesign will make the problem go away.

cptvitamin commented on poll/public/html/survey.html:
```diff
@@ -0,0 +1,55 @@
+                <td class="survey-question">
+                    {% if question.img %}
+                        <div class="poll-image-td">
+                                <img src="{{question.img}}" />
```
_MUST_ have an `alt` attribute

smarnach replied:
Done.

cptvitamin commented on poll/public/html/survey.html:
```diff
@@ -0,0 +1,55 @@
+                </td>
+            {% for answer, answer_details in answers %}
+                <td class="survey-option">
+                    <input type="radio" name="{{key}}" value="{{answer}}"{% if question.choice == answer %} checked{% endif %}/>
```
This input must have a `<label>` associated with it.

smarnach replied:
This is a table, with the labels at the top of each column, and then a grid of radio buttons:

![image](https://cloud.githubusercontent.com/assets/249196/8020174/af886cb2-0c6f-11e5-9424-0f3f07c3f365.png)

Labels can only be associated with a single radio button, so I can't associate the labels in the column headers with the buttons.  What is the best way to deal with this?  Should I add labels to each table cell with a screen reader class?

cptvitamin replied:
Yes something like this would work:
`<label><input type="radio" name="{{key}}" value="{{answer}}"{% if question.choice == answer %} checked{% endif %}/><span class="sr">Yes</span></label>`

smarnach replied:
Done.

cptvitamin commented on poll/public/html/survey.html:
```diff
@@ -0,0 +1,55 @@
+    {% endif %}
+
+    {% if can_view_private_results %}
+        <div class="view-results-button-wrapper"><a class="view-results-button">View results</a></div>
```
Needs to be a button.  This `<a>` has no `href` making it uber bad.

smarnach replied:
Done.

sarina commented:
@nedbat please provide review of xblock architecture

nedbat commented on poll/poll.py:
```diff
@@ -0,0 +1,849 @@
+        Checks to see if the user has permissions to view private results.
+        This only works inside the LMS.
+        """
+        if HAS_EDX_ACCESS and hasattr(self.runtime, 'user') and hasattr(self.runtime, 'course_id'):
```
ouch. (and by that, I mean, it's really too bad we don't have a supported way for you to get this).
But thanks for writing it defensively anyway. :)

nedbat commented on poll/poll.py:
```diff
@@ -0,0 +1,849 @@
+                # Check if user is member of a group that is explicitly granted
+                # permission to view the results through django configuration.
+                group_names = getattr(settings, 'XBLOCK_POLL_EXTRA_VIEW_GROUPS', [])
+                if group_names:
```
The missing-else clause here will return None. Removing more else's could help.

smarnach replied:
Done.

nedbat commented on poll/poll.py:
```diff
@@ -0,0 +1,849 @@
+        """
+        Find out if any answer has an image, since it affects layout.
+        """
+        return any(value['img'] for value in dict(field).values())
```
A small point: there are many places where the field is converted from lists to dicts, etc.  This could be avoided here with:
`any(value['img'] for key, value in field)`

smarnach replied:
Done.

nedbat commented on poll/poll.py:
```diff
@@ -0,0 +1,849 @@
+        we just clean it up on first access within the LMS, in case the studio
+        has made changes to the answers.
+        """
+        answers = OrderedDict(self.answers)
```
nit: no need for an ordereddict here.

smarnach replied:
Done.

nedbat commented on poll/poll.py:
```diff
@@ -0,0 +1,849 @@
+
+        context.update({
+            'choice': choice,
+            # Offset so choices will always be True.
```
I'm not sure what this offset refers to?

e-kolpakov replied:
@nedbat [stale comment](https://github.com/open-craft/xblock-poll/commit/f56c62d460d5c061605f9d711cc1991de8990ede#diff-39725552b3964905d716e989c7ca4e23R86) - thank you for noticing, I'll remove it

nedbat commented on poll/poll.py:
```diff
@@ -0,0 +1,849 @@
+
+    def student_view(self, context=None):
+        """
+        The primary view of the PollBlock, shown to students
```
micro-nit: this is the view of the SurveyBlock.

smarnach replied:
Done.

nedbat commented:
The Python code looks good to me: :+1:.  It's too bad all the scope-required recalculations on the student_summary scoped data, but that's the state of things now.

sarina commented:
@e-kolpakov  @antoviaque : let me know when you have responses to Ned & Mark's comments in your follow-up PR and we'll work on getting the review completed!

Quick question: does this block emit analytics events? If not, it should - the events should be very simple, no more than a general "problem-check" event that capa problems emit in edx-platform. Goal should be that you should be able to follow a learner's analytics event stream and understand their engagement with the course.

@antoviaque how do you want to do the blog post? I think we should work on getting a draft put together, so we can publish something to our blog and send to the edx-code mailing list once this is available.

antoviaque commented:
@sarina Thanks! For the blog post, it's a good question, McKinsey didn't reply to my last email about this -- I would expect them to contribute to the writing, so I'll ask them again. And the plan looks good to me.

sarina commented:
@antoviaque I spoke with Leslie about this block and she loves it (!!!) but has one bit of feedback + a question:

Only feedback I have is that the authoring interface isn't great. I would propose to them doing what we did in ORA (screenshot below), which is create three tabs in the XBlock editor for: 

Question
Answers
Settings 

instead of what you have now, which is that weird "Add question", "add answer" across the bottom. My sense is that this shouldn't be too much work, given that they already custom-configured the "add question" and "add answer" links. 

![image 6](https://cloud.githubusercontent.com/assets/1985317/7846528/13eba2d2-048c-11e5-9c5f-78f216b595ac.png)

How feasible would that change be?

antoviaque commented:
@sarina That looks like a nice change! I'll need to run this by the McKinsey team though, as we're getting close to the time budget cap we had on this so they might need to approve an extension, and I'd like to make sure they are ok with the UI change. 

I also still owe you an answer about the blog - I haven't forgotten : )

I'll let you know asap on both points.

sarina commented:
@antoviaque sure thing - let me know either way.

sarina commented on README.md:
```diff
@@ -0,0 +1,239 @@
+An analytics event will not be fired upon the student viewing the results, as the results are never visible. A user
+will see a thank you message and any feedback provided upon submission:
+
+![Private Results Submission](doc_img/private_results_submission.png)
```
I don't understand how, as an instructor, to get the results of a private survey... am I missing something here?

antoviaque replied:
It is displayed to course staff - see https://github.com/mckinseyacademy/xblock-poll/pull/13/files#diff-39725552b3964905d716e989c7ca4e23R186

sarina replied:
I'm looking at the sandbox and I don't understand how, as the instructor, I can get the results from my block.

Here is where I am: http://sandbox3.opencraft.com/courses/TestOrg/ImportCourse1/now/courseware/d5365b39b81348fb804bf57feaff4264/25bc82de92884537a7fba02fd31f8bc0/

I am logged in as staff. I looked at Studio and answered the question through all the resubmits, thinking maybe one of those options would show me the results of the poll, but neither works.

Either the UX needs to be clearer, or the README needs to explicitly explain - right here! - how instructors can see the results.

If you can tell me how I can, from the sandbox, see the results of the poll as staff I can help suggest how to word documentation.

antoviaque replied:
@sarina Hum, true, I can't see the results either with the staff account - a "View results" link should appear right below the survey with private results enabled, to allow to display them, but that doesn't seem to be working. I'll let @smarnach look at this when he works on the review comments, it looks like there could be a bug there - rights might be setup slightly differently on the instance where we currently run this. And +1 to improving the doc here, in any case.

smarnach replied:
This bug should be fixed now (the upload to the sandbox is still pending).

antoviaque commented:
For the interface change, after discussion with the client, all good (with a time cap of 5 hours on that change).

Btw @smarnach is taking over addressing the review comments here, as @e-kolpakov is completing other work before going on holidays.

smarnach commented:
@sarina You asked about analytics events a while ago.  We already emit a few events, see [the documentation in the README](https://github.com/mckinseyacademy/xblock-poll#analytics).  Is this enough, or do we need anything else?

smarnach commented:
@cptvitamin I addressed your notes, but have a few questions.  Could you please take a look?

sarina commented:
@smarnach great, thanks for pointing that out. I think those events are sufficient.

sarina commented on poll/public/handlebars/poll_results.handlebars:
```diff
@@ -0,0 +1,43 @@
+    {{/each}}
+    </ul>
+    <input class="input-main" type="button" name="poll-submit" value="Submit" disabled>
+    <div class="poll-footnote">Results gathered from {{total}} respondent{{#if plural}}s{{/if}}.</div>
```
ugh we need i18n support, unfortunately we can't have in xblocks. sigh. not worth fixing but initially you see, "from 0 respondent", because English counts 0 as a plural.

lamagnifica commented:
@smarnach, hello, Alison here from the edX doc team. I'm looking at the event samples in the readme, getting a draft together about the 4 new events for the Research guide, and I have just a question or two for you. 
In the `xblock.survey.submitted` event type, can you help me understand where the name::value members of the `event.choices` object get set? I see that the example event doesn't correlate exactly to the earlier example of a survey, so I am wondering if the complete label defined in Studio for each question and answer will be supplied. 
That is, for the example survey here https://github.com/mckinseyacademy/xblock-poll#survey-examples, would the `event.choices` object look something like this?
"choices": {
  "I feel like this test will pass.": "Strongly Agree",
  "I like testing software": "Agree",
  "Testing is not necessary": "Disagree",
  "I would fake a test result to get software deployed.": "Strongly Disagree"
}
or is some truncation done, or is there another source of the "one word": "one letter" name::value format shown in the sample event?
Similarly, in the `xblock.poll.submitted` `event.choice` field, is it safe to assume that each answer defined in Studio is given an internal identifying letter, from A to Z? I see in the readme that the order of the responses is important, and that the arrow controls should be used to change the display order of the possible answers. Is there any way to tell what answer text corresponds to an assigned letter after this type of resequencing takes place?

smarnach commented:
@lamagnifica Sorry for the delay.  I didn't know the answers to your question, so I had to investigate this myself before answering.

Regarding the xblock.survey.submitted event, the choice questions and answers are exactly like you enter them in Studio, and the choices object for the example you quoted will look the way you presumed it to be.  There is no other shortened form.

Regarding the assignment of choice keys for polls, that's a bit more involved.  The keys aren't guaranteed to be single letters, but can rather be identifying strings.  They are generated from the label or image when you first create an answer, and are kept the same from this point, regardless of any reordering etc.  I'm not sure whether there's an easy way to find out the current set of keys for a poll.  @Kelketek Would you be able to shed more light on this question?

Kelketek commented:
@smarnach The events are fired in the Python. You should be able to pull the options in as a dictionary like is done elsewhere in the code, since it's an ordered list of key, value pairs, and that way get the options. I don't believe the event is actually that clear in its output. I think I made it to where it just spits out whatever the key of the answer was, since I was not entirely sure what the design goals of an event should have been at the time and I figured there'd be some other way to look up what those keys meant if doing an analysis. Unless you've changed it to give more detail?

smarnach commented:
@Kelketek Thanks.  I've found the code sending the event, but the point I'm not completely clear about is how keys are assigned to choices for polls, and how course authors could find out what key belongs to what choice, which is what they need to do to make sense of the `xblock.poll.submitted` events.  I'm not trying to change anything in the code, but rather to understand what it is currently doing to help Alison with the documentation.

lamagnifica commented:
Hello @Kelketek and @smarnach, I met briefly with @shnayder, the product manager for analytics at edX. The goal of the `xblock.poll.submitted` event would ideally be to give researchers (now) and course teams using Insights (future) data about the selections that students make when polled. Providing some info about how they can map what's in the `event.choice` field to the poll options that are defined in Studio would be helpful if you can provide it without heroic effort. Going forward, we might want to change the content or derivation of the `event.choice` field to provide better/easier visibility into this data. 
@sarina, @shnayder doesn't see this line of investigation as blocking this PR from merging.

smarnach commented:
@lamagnifica The `event.choice` field contains an internal "key" for the choice.  As I said before, the key is auto-assigned by the code based on the label or the image, and I don't know of a way to see the keys for each choice, and I don't completely understand the way the keys are generated if you e.g. only provide an image.  I can try and find out more tomorrow by instrumenting the code and playing with a few examples.

This particular PR isn't meant to be merged – it only exists for review.  The PR the comments are addressed in is already merged, so this discussion isn't blocking anything, and we can of course make changes to the event in a different PR later.

sarina commented:
The X block is already merged to edx platform and will be included in the
next release.
On Jul 1, 2015 4:27 PM, "Sven Marnach" notifications@github.com wrote:

> @lamagnifica https://github.com/lamagnifica The event.choice field
> contains an internal "key" for the choice. As I said before, the key is
> auto-assigned by the code based on the label or the image, and I don't know
> of a way to see the keys for each choice, and I don't completely understand
> the way the keys are generated if you e.g. only provide an image. I can try
> and find out more tomorrow by instrumenting the code and playing with a few
> examples.
> 
> This particular PR isn't meant to be merged – it only exists for review.
> The PR the comments are addressed in is already merged, so this discussion
> isn't blocking anything, and we can of course make changes to the event in
> a different PR later.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/mckinseyacademy/xblock-poll/pull/13#issuecomment-117816220
> .

Kelketek committed:
Update test environment.

Kelketek committed:
Handle multi-line poll answers more sanely.

Kelketek committed:
Merge pull request #16 from mckinseyacademy/multiline

Fixes for Multiline poll questions.

Kelketek committed:
Update README.md

antoviaque committed:
Merge pull request #17 from mckinseyacademy/Kelketek-patch-1

Notification to make PRs against the open-craft fork.
